File: code/dynamic_programming/src/shortest_common_supersequence/shortest_common_supersequence.cpp
/* Part of Cosmos by OpenGenus Foundation */

#include <iostream>
#include <cstring>

using namespace std;

const int MAX = 1010;

int memo[MAX][MAX]; // used for memoization

// Function to find length of shortest common supersequence
// between sequences X[0..m-1] and Y[0..n-1]
int SCSLength(string& X, string& Y, int m, int n)
{
    // if we have reached the end of either sequence, return
    // length of other sequence
    if (m == 0 || n == 0)
        return n + m;

    // if we have already computed this state
    if (memo[m - 1][n - 1] > -1)
        return memo[m - 1][n - 1];

    // if last character of X and Y matches
    if (X[m - 1] == Y[n - 1])
        return memo[m - 1][n - 1] = SCSLength(X, Y, m - 1, n - 1) + 1;
    else
        // else if last character of X and Y don't match
        return memo[m - 1][n - 1] = 1 + min(SCSLength(X, Y, m, n - 1),
                                            SCSLength(X, Y, m - 1, n));
}

// helper function
int findSCSLength(string& X, string& Y)
{
    memset(memo, -1, sizeof memo);
    return SCSLength(X, Y, X.length(), Y.length());
}

// main function
int main()
{
    string X = "ABCBDAB", Y = "BDCABA";

    cout << "The length of shortest common supersequence is "
         << findSCSLength(X, Y) << '\n';

    return 0;
}

File: code/dynamic_programming/src/shortest_common_supersequence/shortest_common_supersequence.py
# Part of Cosmos by OpenGenus Foundation

# shortest common supersequence using dp
def min(a, b):
    if a > b:
        return b
    else:
        return a


def superSeq(X, Y, m, n):
    dp = []
    for i in range(m + 1):
        l2 = []
        for j in range(n + 1):
            if not (i):
                l2.append(j)

            elif not (j):
                l2.append(i)

            elif X[i - 1] == Y[j - 1]:
                l2.append(1 + dp[i - 1][j - 1])

            else:
                l2.append(1 + min(dp[i - 1][j], l2[j - 1]))
        dp.append(l2)
    return dp[m][n]


X = "FABCDF"
Y = "HBABCDEF"
# longest supersequence can be HBFABCDEF
print("Length of the shortest supersequence = ", superSeq(X, Y, len(X), len(Y)))
